File: algorithms/python/_167_TwoSumII_InputArrayIsSorted/_167_TwoSumII_InputArrayIsSorted.py
# two-pointer
def twoSum1(self, numbers, target):
    l, r = 0, len(numbers) - 1
    while l < r:
        s = numbers[l] + numbers[r]
        if s == target:
            return [l + 1, r + 1]
        elif s < target:
            l += 1
        else:
            r -= 1


# dictionary
def twoSum2(self, numbers, target):
    dic = {}
    for i, num in enumerate(numbers):
        if target - num in dic:
            return [dic[target - num] + 1, i + 1]
        dic[num] = i


# binary search
def twoSum(self, numbers, target):
    for i in range(len(numbers)):
        l, r = i + 1, len(numbers) - 1
        tmp = target - numbers[i]
        while l <= r:
            mid = l + (r - l) // 2
            if numbers[mid] == tmp:
                return [i + 1, mid + 1]
            elif numbers[mid] < tmp:
                l = mid + 1
            else:
                r = mid - 1


# Here's a summary of how long it takes to run for three methods:
#
# Two pointers: 43 ms
# Dictionary: 46 ms
# Binary Search: 75 ms

# It was odd to me that binary search is significantly slower than other two methods;
# it should be faster since we are using the sorted feature.
# It turns out that if we don't repeat investigating the elements that have been already investigated,
# binary search becomes the fastest method.
# Here's the modified version which takes 35 ms:

def twoSum(self, numbers, target):
    investigatedSoFar = []
    for i in range(len(numbers)):
        if not numbers[i] in investigatedSoFar:
            investigatedSoFar.append(numbers[i])
            l, r = i + 1, len(numbers) - 1
            tmp = target - numbers[i]
            while l <= r:
                mid = l + (r - l) // 2
                if numbers[mid] == tmp:
                    return ([i + 1, mid + 1])
                elif numbers[mid] < tmp:
                    l = mid + 1
                else:
                    r = mid - 1


# For the two-pointer one, this seems to be faster

def twoSum1(self, numbers, target):
    f, l = 0, len(numbers) - 1
    while numbers[f] + numbers[l] != target:
        if numbers[f] + numbers[l] > target:
            l -= 1
        else:
            f += 1
    return [f + 1, l + 1]

File: algorithms/java/src/_167_TwoSumII_InputArrayIsSorted/_167_TwoSumII_InputArrayIsSorted.java
package _167_TwoSumII_InputArrayIsSorted;

public class _167_TwoSumII_InputArrayIsSorted {
    public static void main(String[] args) {
        _167_TwoSumII_InputArrayIsSorted_ so = new _167_TwoSumII_InputArrayIsSorted_();

    }
}

class _167_TwoSumII_InputArrayIsSorted_ {
    public int[] twoSum(int[] num, int target) {
        int[] indice = new int[2];
        if (num == null || num.length < 2) return indice;
        int left = 0, right = num.length - 1;
        while (left < right) {
            int v = num[left] + num[right];
            if (v == target) {
                indice[0] = left + 1;
                indice[1] = right + 1;
                break;
            } else if (v > target) {
                right--;
            } else {
                left++;
            }
        }
        return indice;
    }

}

class _167_TwoSumII_InputArrayIsSorted_1 {
    public int[] twoSum(int[] numbers, int target) {
        if (numbers == null || numbers.length == 0) {
            return new int[2];
        }
        int start = 0;
        int end = numbers.length - 1;
        while (start < end) {
            if (numbers[start] + numbers[end] == target) {
                return new int[]{start + 1, end + 1};
            } else if (numbers[start] + numbers[end] > target) {
                // move end forward to the last value that numbers[end] <= target - numbers[start]
                end = largestSmallerOrLastEqual(numbers, start, end, target - numbers[start]);
            } else {
                // move start backword to the first value that numbers[start] >= target - numbers[end]
                start = smallestLargerOrFirstEqual(numbers, start, end, target - numbers[end]);
            }
        }
        return new int[2];
    }

    private int largestSmallerOrLastEqual(int[] numbers, int start, int end, int target) {
        int left = start;
        int right = end;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (numbers[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return right;
    }

    private int smallestLargerOrFirstEqual(int[] numbers, int start, int end, int target) {
        int left = start;
        int right = end;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (numbers[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }
}
